pkg/kube/configmapwatcher/configmapwatcher.go

```diff
@@ -0,0 +1,137 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package configmapwatcher
+
+import (
+	"fmt"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	infomersv1 "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"istio.io/istio/pkg/kube"
+	"istio.io/pkg/log"
+)
+
+// Controller watches a ConfigMap and calls the given callback when the ConfigMap changes.
+type Controller struct {
+	client   kube.Client
+	informer infomersv1.ConfigMapInformer
+	queue    workqueue.RateLimitingInterface
+
+	configMapNamespace string
+	configMapName      string
+	callback           func()
```

Comment:
nit: maybe it should return the configmap? otherwise they need to call .Get on the callback which is a bit clunky.

Reply:
That makes sense. Originally I was thinking I'd want to do a synchronous read as well so calling .Get would already also be done outside of the callback anyway. But maybe all reads would be done through the callback since it makes sense for the cache to sync before a first read, which would trigger the callback. I think it would make sense to return the ConfigMap if we keep the queue, because errors for reading the ConfigMap would need to be processed within the controller, and the queue helps with dealing with those

Reply:
I do not think anyone needs to read the configmap (unless they want to call .Get() instead of the callback). The update/add/delete handlers already pass the object, so we shouldn't need to fetch it again?

Reply:
Yeah that's what I meant by the callback being triggered. Maybe after doing all the other code I can refactor this